README.rst
pycode_similar
==============

This is a simple plagiarism detection tool for python code, the basic idea is to normalize python AST representation and use difflib to get the modification from referenced code to candidate code. The plagiarism defined in pycode_similar is how many referenced code is plagiarized by candidate code, which means swap referenced code and candidate code will get different result.

It only cost me a couple of hours to implement this tool, so there is still a long way to improve the speed and accuracy, but it already performs great in detecting the plagiarism of new recruits' homeworks in our company.

Compare to Moss
---------------

- pure python implementation
- only contains one source file
- no third-party dependency (except `zss  <https://pypi.python.org/pypi/zss>`_ when use TreeDiff)
- no need to register account for Moss
- no need of network to access Moss

This tool was born before I know there is a `Moss (for a Measure Of Software Similarity)  <https://theory.stanford.edu/~aiken/moss/>`_ to determine the similarity of programs. And I have tried many ways to register account for Stanford Moss, but still can't get a valid account. So, I have no accurate comparison between pycode_similar and Moss.

Installation
--------------

If you don't have much time, just perform

 `$ pip install pycode_similar`

which will install the module(without tests) on your system.

Also, you can just copy & paste the pycode_similar.py which require no third-party dependency.


Usage
--------------

Just use it as a standard command line tool if pip install properly.

.. code-block:: text

	$ pycode_similar
	usage: pycode_similar [-h] [-l L] [-p P] [-k] [-m] files files

	A simple plagiarism detection tool for python code

	positional arguments:
	  files       the input files

	optional arguments:
	  -h, --help          show this help message and exit
	  -l L                if AST line of the function >= value then output detail (default: 4)
	  -p P                if plagiarism percentage of the function >= value then output detail (default: 0.5)
	  -k, --keep-prints   keep print nodes
	  -m, --module-level  process module level nodes

	pycode_similar: error: too few arguments

Of course, you can use it as a python library, too.

.. code-block:: python

	import pycode_similar
	pycode_similar.detect([referenced_code_str, candidate_code_str1, candidate_code_str2, ...], diff_method=pycode_similar.UnifiedDiff, keep_prints=False, module_level=False)
	
	
Implementation
--------------
This tool has implemented two diff methods: line based diff(UnifiedDiff) and tree edit distance based diff(TreeDiff), both of them are run in function AST level.

- UnifiedDiff, diff normalized function AST string lines, naive but efficiency.
- TreeDiff, diff function AST, very slow and the result is not good for small functions. (depends on `zss  <https://pypi.python.org/pypi/zss>`_)

So, when run this tool in cmd, the default diff method is UnifiedDiff. And you can switch to TreeDiff when use it as a library.


Testing
--------------
If you have the source code you can run the tests with

 `$ python pycode_similar/tests/test_cases.py`
 
Or perform

.. code-block:: text

	$ python pycode_similar.py pycode_similar/tests/original_version.py pycode_similar.py

	ref: tests/original_version.py
	candidate: pycode_similar.py
	80.14 % (803/1002) of ref code structure is plagiarized by candidate.
	candidate function plagiarism details (AST lines >= 4 and plagiarism percentage >= 0.5):
	1.0 : ref FuncNodeCollector._mark_docstring_sub_nodes<24:4>, candidate FuncNodeCollector._mark_docstring_sub_nodes<27:4>
	1.0 : ref FuncNodeCollector._mark_docstring_nodes<54:8>, candidate FuncNodeCollector._mark_docstring_nodes<57:8>
	1.0 : ref FuncNodeCollector.generic_visit<69:4>, candidate FuncNodeCollector.generic_visit<72:4>
	1.0 : ref FuncNodeCollector.visit_Str<74:4>, candidate FuncNodeCollector.visit_Str<78:4>
	1.0 : ref FuncNodeCollector.visit_Name<83:4>, candidate FuncNodeCollector.visit_Name<88:4>
	1.0 : ref FuncNodeCollector.visit_Attribute<89:4>, candidate FuncNodeCollector.visit_Name<88:4>
	1.0 : ref FuncNodeCollector.visit_ClassDef<95:4>, candidate FuncNodeCollector.visit_ClassDef<100:4>
	1.0 : ref FuncNodeCollector.visit_FunctionDef<101:4>, candidate FuncNodeCollector.visit_FunctionDef<106:4>
	1.0 : ref FuncInfo.__init__<141:4>, candidate FuncInfo.__init__<161:4>
	1.0 : ref FuncInfo.__str__<151:4>, candidate FuncInfo.__str__<171:4>
	1.0 : ref FuncInfo.func_code<162:4>, candidate FuncInfo.func_code<182:4>
	1.0 : ref FuncInfo.func_code_lines<168:4>, candidate FuncInfo.func_code_lines<188:4>
	1.0 : ref FuncInfo.func_ast<174:4>, candidate FuncInfo.func_ast<194:4>
	1.0 : ref FuncInfo.func_ast_lines<180:4>, candidate FuncInfo.func_ast_lines<200:4>
	1.0 : ref FuncInfo._retrieve_func_code_lines<186:4>, candidate FuncInfo._retrieve_func_code_lines<206:4>
	1.0 : ref FuncInfo._iter_node<208:4>, candidate FuncInfo._iter_node<228:4>
	1.0 : ref FuncInfo._dump<232:4>, candidate FuncInfo._dump<252:4>
	1.0 : ref FuncInfo._inner_dump<242:8>, candidate FuncInfo._inner_dump<262:8>
	1.0 : ref ArgParser.error<267:4>, candidate ArgParser.error<291:4>
	0.95: ref unified_diff<281:0>, candidate UnifiedDiff._gen<339:8>
	0.92: ref FuncNodeCollector.__init__<18:4>, candidate FuncNodeCollector.__init__<20:4>
	0.92: ref FuncNodeCollector.visit_Compare<108:4>, candidate FuncNodeCollector._simple_nomalize<117:8>
	0.89: ref FuncNodeCollector.visit_Expr<79:4>, candidate FuncNodeCollector.visit_Expr<83:4>
	
Click `here  <https://github.com/fyrestone/pycode_similar/commit/149182beee460cbaf21d0995aa442a079ddf1fa9#diff-a30b425e81348c978616747430632fa8>`_
to view this diff -> `0.92: ref FuncNodeCollector.visit_Compare<108:4>, candidate FuncNodeCollector._simple_nomalize<117:8>`

Repository
--------------

The project is hosted on GitHub. You can look at the source here:

 https://github.com/fyrestone/pycode_similar
 


CHANGES.md
HISTORY
--------

## 1.4 (2020-8-8)
- Add the `--keep-prints` option for keeping print nodes, default is False. (By @thektulu)
- Add the `--module-level` option for module level diff, default is False. (By @thektulu)

## 1.3 (2019-12-8)
- Fix python 3.8 compatibility issue [#4](https://github.com/fyrestone/pycode_similar/issues/4) (Thanks to aginiewicz)
- Remove import expr

## 1.2 (2018-1-18)

- Fix python3 compatibility issue.
- Remove print for python23.

## 1.1 (2017-10-04)

- Add TreeDiff method based on zss.
- Fixed sub nodes of Expr are not normalized.
- Fixed incorrect result when detect target contains no function.

## 1.0 (2017-10-01)

- Init


pycode_similar/tests/test_cases.py
import os
import sys

sys.path.insert(0, os.path.realpath(os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))))

import unittest
import pycode_similar


class TestCases(unittest.TestCase):

    def test_basic_detect(self):
        s1 = """
def foo(a):
    if a > 1:
        return True
    return False
        """
        s2 = """
class A(object):
    def __init__(self, a):
        self._a = a
        
    def bar(self):
        if self._a > 2:
            return True
        return False
        """
        result = pycode_similar.detect([s1, s2])
        self.assertGreater(result[0][1][0].plagiarism_percent, 0.5)

    def test_name(self):
        s1 = """
def foo(a):
    if a > 1:
        return True
    return False
            """
        s2 = """
def bar(b):
    if b > 1:
        return True
    return False
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_equal(self):
        s1 = """
def foo(a):
    if a == 1:
        return True
    return False
            """
        s2 = """
def bar(b):
    if 1 == b:
        return True
    return False
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_gt_lt(self):
        s1 = """
def foo(a):
    if a > 1:
        return True
    return False
            """
        s2 = """
def bar(b):
    if 1 < b:
        return True
    return False
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_gte_lte(self):
        s1 = """
def foo(a):
    if a >= 1:
        return True
    return False
            """
        s2 = """
def bar(b):
    if 1 <= b:
        return True
    return False
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_space_and_comments(self):
        s1 = """
def foo(a):
    \"""
    foo comments.
    \"""
    if a >= 1:
        return True
        
    # this should return False    
    return False
            """
        s2 = """
def bar(b):
# bar comments.
    if 1 <= b:
        \"""
        This should
        return True
        \"""
        return True
    return False
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_expr(self):
        s1 = """
def foo(a):
    yield c
            """
        s2 = """
def bar(b):
    yield a
            """
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_no_function(self):
        s1 = """
def foo(a):
    c = a
            """
        s2 = """
class B(object):
    pass
            """
        try:
            result = pycode_similar.detect([s2, s1])
        except Exception as ex:
            self.assertEqual(ex.source, 0)
        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 0)

    def test_strip_print(self):
        s1 = """
def foo(a):
    a = b
    print('abc', a)
            """
        s2 = """
def foo(a):
    print('abc', bar())
    a = b
            """

        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_strip_import(self):
        s1 = """
def foo():
    import sys
    from os import path
            """
        s2 = """
def foo():
    import os
    import ast
    from collections import Counter
            """

        result = pycode_similar.detect([s1, s2])
        self.assertEqual(result[0][1][0].plagiarism_percent, 1)

    def test_module_level(self):
        s1 = """
def main():
    s = 0
    for j in range(10):
        for i in range(10):
            if i > j:
                s += i + j
    print(s)

if __name__ == '__main__':
    main()
"""
        s2 = """
s = 0
for j in range(10):
    for i in range(10):
        if i > j:
            s += i + j
print(s)
"""
        result = pycode_similar.detect([s1, s2], module_level=True)
        sum_plagiarism_percent, *tail = pycode_similar.summarize(result[0][1])
        # s1.main vs s2.__main__ AND s1.__main__ vs s2.__main__
        self.assertGreater(sum_plagiarism_percent, 0.6)

        result = pycode_similar.detect([s2, s1], module_level=True)
        sum_plagiarism_percent, *tail = pycode_similar.summarize(result[0][1])
        # s2.__main__ vs s1.main
        self.assertGreater(sum_plagiarism_percent, 0.85)

    def test_keep_prints(self):
        s1 = """
a = []
for j in range(10):
    for i in range(10):
        a.append(i - j)
print(a)
print(abs(el) * el for el in a if abs(el) > 2)
"""
        s2 = """
a = [
    i - j
    for j in range(10)
    for i in range(10)
]
print(a)
print(abs(el) * el for el in a if abs(el) > 2)
"""

        result = pycode_similar.detect([s1, s2], module_level=True, keep_prints=False)
        self.assertLess(result[0][1][0].plagiarism_percent, 0.2)

        result = pycode_similar.detect([s1, s2], module_level=True, keep_prints=True)
        self.assertGreater(result[0][1][0].plagiarism_percent, 0.5)


if __name__ == "__main__":
    #     import sys;sys.argv = ['', 'Test.test_reload_custom_code_after_changes_in_class']
    unittest.main()


pycode_similar/pycode_similar.py
# -*- coding: utf-8 -*-
__author__ = 'fyrestone@outlook.com'
__version__ = '1.4'

import sys
import ast
import difflib
import operator
import argparse
import itertools
from collections import Counter

# avoid using six to keep dependency clean
if sys.version_info >= (3, 3):
    import collections.abc as collections
else:
    import collections

if sys.version_info[0] == 3:
    string_types = str
else:
    string_types = basestring


class BaseNodeNormalizer(ast.NodeTransformer):
    """
    Clean node attributes, delete the attributes that are not helpful for recognition repetition.
    """

    def __init__(self, keep_prints=False):
        super(BaseNodeNormalizer, self).__init__()
        self.keep_prints = keep_prints
        self._node_count = 0

    @staticmethod
    def _mark_docstring_sub_nodes(node):
        """
        Inspired by ast.get_docstring, mark all docstring sub nodes.

        Case1:
        regular docstring of function/class/module

        Case2:
        def foo(self):
            '''pure string expression'''
            for x in self.contents:
                '''pure string expression'''
                print x
            if self.abc:
                '''pure string expression'''
                pass

        Case3:
        def foo(self):
            if self.abc:
                print('ok')
            else:
                '''pure string expression'''
                pass

        :param node: every ast node
        :return:
        """

        def _mark_docstring_nodes(body):
            if body and isinstance(body, collections.Sequence):
                for n in body:
                    if isinstance(n, ast.Expr) and isinstance(n.value, ast.Str):
                        n.is_docstring = True

        node_body = getattr(node, 'body', None)
        _mark_docstring_nodes(node_body)
        node_orelse = getattr(node, 'orelse', None)
        _mark_docstring_nodes(node_orelse)

    @staticmethod
    def _is_docstring(node):
        return getattr(node, 'is_docstring', False)

    def generic_visit(self, node):
        self._node_count = self._node_count + 1
        self._mark_docstring_sub_nodes(node)
        return super(BaseNodeNormalizer, self).generic_visit(node)

    def visit_Constant(self, node):
        # introduce a special value for erasing constant node value,
        # del node.value will make node.s and node.n raise Exception.
        # for Python 3.8
        dummy_value = '__pycode_similar_dummy_value__'
        if type(node) == str:
            node.value = dummy_value
        self.generic_visit(node)

    def visit_Str(self, node):
        del node.s
        self.generic_visit(node)
        return node

    def visit_Expr(self, node):
        if not self._is_docstring(node):
            self.generic_visit(node)
            if hasattr(node, 'value'):
                return node

    def visit_arg(self, node):
        """
        remove arg name & annotation for python3
        :param node: ast.arg
        :return:
        """
        del node.arg
        del node.annotation
        self.generic_visit(node)
        return node

    def visit_Name(self, node):
        del node.id
        del node.ctx
        self.generic_visit(node)
        return node

    def visit_Attribute(self, node):
        del node.attr
        del node.ctx
        self.generic_visit(node)
        return node

    def visit_Call(self, node):
        func = getattr(node, 'func', None)
        if not self.keep_prints and func and isinstance(func, ast.Name) and func.id == 'print':
            return  # remove print call and its sub nodes for python3
        self.generic_visit(node)
        return node

    def visit_Compare(self, node):

        def _simple_nomalize(*ops_type_names):
            if node.ops and len(node.ops) == 1 and type(node.ops[0]).__name__ in ops_type_names:
                if node.left and node.comparators and len(node.comparators) == 1:
                    left, right = node.left, node.comparators[0]
                    if type(left).__name__ > type(right).__name__:
                        left, right = right, left
                        node.left = left
                        node.comparators = [right]
                        return True
            return False

        if _simple_nomalize('Eq'):
            pass

        if _simple_nomalize('Gt', 'Lt'):
            node.ops = [{ast.Lt: ast.Gt, ast.Gt: ast.Lt}[type(node.ops[0])]()]

        if _simple_nomalize('GtE', 'LtE'):
            node.ops = [{ast.LtE: ast.GtE, ast.GtE: ast.LtE}[type(node.ops[0])]()]

        self.generic_visit(node)
        return node

    def visit_Print(self, node):
        if not self.keep_prints:
            # remove print stmt for python2
            return
        self.generic_visit(node)
        return node

    def visit_Import(self, node):
        # remove import ...
        pass

    def visit_ImportFrom(self, node):
        # remove from ... import ...
        pass


class ModuleNodeCollector(BaseNodeNormalizer):
    """
    Normalize and remove all class nodes and function nodes - leave only module level nodes.
    """

    def __init__(self, *args, **kwargs):
        super(ModuleNodeCollector, self).__init__(*args, **kwargs)
        self._module_node = None

    def visit_ClassDef(self, node):
        # remove class ...
        return

    def visit_FunctionDef(self, node):
        # remove function ...
        return

    def visit_Module(self, node):
        self._module_node = node
        count = self._node_count
        self.generic_visit(node)
        node.name = '__main__'
        node.lineno = 1
        node.col_offset = 0
        node.nsubnodes = self._node_count - count
        return node

    def get_module_node(self):
        return self._module_node


class FuncNodeCollector(BaseNodeNormalizer):
    """
    Normalize and collect all function nodes.
    """

    def __init__(self, *args, **kwargs):
        super(FuncNodeCollector, self).__init__(*args, **kwargs)
        self._curr_class_names = []
        self._func_nodes = []
        self._last_node_lineno = -1

    def generic_visit(self, node):
        self._last_node_lineno = max(getattr(node, 'lineno', -1), self._last_node_lineno)
        return super(FuncNodeCollector, self).generic_visit(node)

    def visit_ClassDef(self, node):
        self._curr_class_names.append(node.name)
        self.generic_visit(node)
        self._curr_class_names.pop()
        return node

    def visit_FunctionDef(self, node):
        node.name = '.'.join(itertools.chain(self._curr_class_names, [node.name]))
        self._func_nodes.append(node)
        count = self._node_count
        self.generic_visit(node)
        node.endlineno = self._last_node_lineno
        node.nsubnodes = self._node_count - count
        return node

    def get_function_nodes(self):
        return self._func_nodes


class FuncInfo(object):
    """
    Part of the astor library for Python AST manipulation.

    License: 3-clause BSD

    Copyright 2012 (c) Patrick Maupin
    Copyright 2013 (c) Berker Peksag

    """

    class NonExistent(object):
        pass

    def __init__(self, func_node, code_lines):
        assert isinstance(func_node, (ast.FunctionDef, ast.Module))
        self._func_node = func_node
        self._code_lines = code_lines
        self._func_name = func_node.__dict__.pop('name', '')
        self._func_code = None
        self._func_code_lines = None
        self._func_ast = None
        self._func_ast_lines = None

    def __str__(self):
        return '<' + type(self).__name__ + ': ' + self.func_name + '>'

    @property
    def func_name(self):
        return self._func_name

    @property
    def func_node(self):
        return self._func_node

    @property
    def func_code(self):
        if self._func_code is None:
            self._func_code = ''.join(self.func_code_lines)
        return self._func_code

    @property
    def func_code_lines(self):
        if self._func_code_lines is None:
            self._func_code_lines = self._retrieve_func_code_lines(self._func_node, self._code_lines)
        return self._func_code_lines

    @property
    def func_ast(self):
        if self._func_ast is None:
            self._func_ast = self._dump(self._func_node)
        return self._func_ast

    @property
    def func_ast_lines(self):
        if self._func_ast_lines is None:
            self._func_ast_lines = self.func_ast.splitlines(True)
        return self._func_ast_lines

    @staticmethod
    def _retrieve_func_code_lines(func_node, code_lines):
        if not isinstance(func_node, (ast.FunctionDef, ast.Module)):
            return []
        if not isinstance(code_lines, collections.Sequence) or isinstance(code_lines, string_types):
            return []
        if getattr(func_node, 'endlineno', -1) < getattr(func_node, 'lineno', 0):
            return []
        lines = code_lines[func_node.lineno - 1: func_node.endlineno]
        if lines:
            padding = lines[0][:-len(lines[0].lstrip())]
            stripped_lines = []
            for l in lines:
                if l.startswith(padding):
                    stripped_lines.append(l[len(padding):])
                else:
                    stripped_lines = []
                    break
            if stripped_lines:
                return stripped_lines
        return lines

    @staticmethod
    def _iter_node(node, name='', missing=NonExistent):
        """Iterates over an object:

           - If the object has a _fields attribute,
             it gets attributes in the order of this
             and returns name, value pairs.

           - Otherwise, if the object is a list instance,
             it returns name, value pairs for each item
             in the list, where the name is passed into
             this function (defaults to blank).

        """
        fields = getattr(node, '_fields', None)
        if fields is not None:
            for name in fields:
                value = getattr(node, name, missing)
                if value is not missing:
                    yield value, name
        elif isinstance(node, list):
            for value in node:
                yield value, name

    @staticmethod
    def _dump(node, name=None, initial_indent='', indentation='    ',
              maxline=120, maxmerged=80, special=ast.AST):
        """Dumps an AST or similar structure:

           - Pretty-prints with indentation
           - Doesn't print line/column/ctx info

        """

        def _inner_dump(node, name=None, indent=''):
            level = indent + indentation
            name = name and name + '=' or ''
            values = list(FuncInfo._iter_node(node))
            if isinstance(node, list):
                prefix, suffix = '%s[' % name, ']'
            elif values:
                prefix, suffix = '%s%s(' % (name, type(node).__name__), ')'
            elif isinstance(node, special):
                prefix, suffix = name + type(node).__name__, ''
            else:
                return '%s%s' % (name, repr(node))
            node = [_inner_dump(a, b, level) for a, b in values if b != 'ctx']
            oneline = '%s%s%s' % (prefix, ', '.join(node), suffix)
            if len(oneline) + len(indent) < maxline:
                return '%s' % oneline
            if node and len(prefix) + len(node[0]) < maxmerged:
                prefix = '%s%s,' % (prefix, node.pop(0))
            node = (',\n%s' % level).join(node).lstrip()
            return '%s\n%s%s%s' % (prefix, level, node, suffix)

        return _inner_dump(node, name, initial_indent)


class ArgParser(argparse.ArgumentParser):
    """
    A simple ArgumentParser to print help when got error.
    """

    def error(self, message):
        self.print_help()
        from gettext import gettext as _

        self.exit(2, _('\n%s: error: %s\n') % (self.prog, message))


class FuncDiffInfo(object):
    """
    An object stores the result of candidate python code compared to referenced python code.
    """

    info_ref = None
    info_candidate = None
    plagiarism_count = 0
    total_count = 0

    @property
    def plagiarism_percent(self):
        return 0 if self.total_count == 0 else (self.plagiarism_count / float(self.total_count))

    def __str__(self):
        if isinstance(self.info_ref, FuncInfo) and isinstance(self.info_candidate, FuncInfo):
            return '{:<4.2}: ref {}, candidate {}'.format(self.plagiarism_percent,
                                                          self.info_ref.func_name + '<' + str(
                                                                  self.info_ref.func_node.lineno) + ':' + str(
                                                                  self.info_ref.func_node.col_offset) + '>',
                                                          self.info_candidate.func_name + '<' + str(
                                                                  self.info_candidate.func_node.lineno) + ':' + str(
                                                                  self.info_candidate.func_node.col_offset) + '>')
        return '{:<4.2}: ref {}, candidate {}'.format(0, None, None)


class UnifiedDiff(object):
    """
    Line diff algorithm to formatted AST string lines, naive but efficiency, result is good enough.
    """

    @staticmethod
    def diff(a, b):
        """
        Simpler and faster implementation of difflib.unified_diff.
        """
        assert a is not None
        assert b is not None
        a = a.func_ast_lines
        b = b.func_ast_lines

        def _gen():
            for group in difflib.SequenceMatcher(None, a, b).get_grouped_opcodes(0):
                for tag, i1, i2, j1, j2 in group:
                    if tag == 'equal':
                        for line in a[i1:i2]:
                            yield ''
                        continue
                    if tag in ('replace', 'delete'):
                        for line in a[i1:i2]:
                            yield '-'
                    if tag in ('replace', 'insert'):
                        for line in b[j1:j2]:
                            yield '+'

        return Counter(_gen())['-']

    @staticmethod
    def total(a, b):
        assert a is not None  # b may be None
        return len(a.func_ast_lines)


class TreeDiff(object):
    """
    Tree edit distance algorithm to AST, very slow and the result is not good for small functions.
    """

    @staticmethod
    def diff(a, b):
        assert a is not None
        assert b is not None

        def _str_dist(i, j):
            return 0 if i == j else 1

        def _get_label(n):
            return type(n).__name__

        def _get_children(n):
            if not hasattr(n, 'children'):
                n.children = list(ast.iter_child_nodes(n))
            return n.children

        import zss
        res = zss.distance(a.func_node, b.func_node, _get_children,
                           lambda node: 0,  # insert cost
                           lambda node: _str_dist(_get_label(node), ''),  # remove cost
                           lambda _a, _b: _str_dist(_get_label(_a), _get_label(_b)), )  # update cost
        return res

    @staticmethod
    def total(a, b):
        #  The count of AST nodes in referenced function
        assert a is not None  # b may be None
        return a.func_node.nsubnodes


class NoFuncException(Exception):
    def __init__(self, source):
        super(NoFuncException, self).__init__('Can not find any functions from code, index = {}'.format(source))
        self.source = source


def detect(pycode_string_list, diff_method=UnifiedDiff, keep_prints=False, module_level=False):
    if len(pycode_string_list) < 2:
        return []

    func_info_list = []
    for index, code_str in enumerate(pycode_string_list):
        root_node = ast.parse(code_str)
        collector = FuncNodeCollector(keep_prints=keep_prints)
        collector.visit(root_node)
        code_utf8_lines = code_str.splitlines(True)
        func_info = [FuncInfo(n, code_utf8_lines) for n in collector.get_function_nodes()]
        if module_level:
            root_node = ast.parse(code_str)
            collector = ModuleNodeCollector(keep_prints=keep_prints)
            collector.visit(root_node)
            module_node = collector.get_module_node()
            module_node.endlineno = len(code_utf8_lines)
            module_info = FuncInfo(module_node, code_utf8_lines)
            func_info.append(module_info)
        func_info_list.append((index, func_info))

    ast_diff_result = []
    index_ref, func_info_ref = func_info_list[0]
    if len(func_info_ref) == 0:
        raise NoFuncException(index_ref)

    for index_candidate, func_info_candidate in func_info_list[1:]:
        func_ast_diff_list = []

        for fi1 in func_info_ref:
            min_diff_value = int((1 << 31) - 1)
            min_diff_func_info = None
            for fi2 in func_info_candidate:
                dv = diff_method.diff(fi1, fi2)
                if dv < min_diff_value:
                    min_diff_value = dv
                    min_diff_func_info = fi2
                if dv == 0:  # entire function structure is plagiarized by candidate
                    break

            func_diff_info = FuncDiffInfo()
            func_diff_info.info_ref = fi1
            func_diff_info.info_candidate = min_diff_func_info
            func_diff_info.total_count = diff_method.total(fi1, min_diff_func_info)
            func_diff_info.plagiarism_count = func_diff_info.total_count - min_diff_value if min_diff_func_info else 0
            func_ast_diff_list.append(func_diff_info)
        func_ast_diff_list.sort(key=operator.attrgetter('plagiarism_percent'), reverse=True)
        ast_diff_result.append((index_candidate, func_ast_diff_list))

    return ast_diff_result


def _profile(fn):
    """
    A simple profile decorator
    :param fn: target function to be profiled
    :return: The wrapper function
    """
    import functools
    import cProfile

    @functools.wraps(fn)
    def _wrapper(*args, **kwargs):
        pr = cProfile.Profile()
        pr.enable()
        res = fn(*args, **kwargs)
        pr.disable()
        pr.print_stats('cumulative')
        return res

    return _wrapper


def summarize(func_ast_diff_list):
    sum_total_count = sum(func_diff_info.total_count for func_diff_info in func_ast_diff_list)
    sum_plagiarism_count = sum(func_diff_info.plagiarism_count for func_diff_info in func_ast_diff_list)
    if sum_total_count == 0:
        sum_plagiarism_percent = 0
    else:
        sum_plagiarism_percent = sum_plagiarism_count / float(sum_total_count)
    return sum_plagiarism_percent, sum_plagiarism_count, sum_total_count


# @_profile
def main():
    """
    The console_scripts Entry Point in setup.py
    """

    def check_line_limit(value):
        ivalue = int(value)
        if ivalue < 0:
            raise argparse.ArgumentTypeError("%s is an invalid line limit" % value)
        return ivalue

    def check_percentage_limit(value):
        ivalue = float(value)
        if ivalue < 0:
            raise argparse.ArgumentTypeError("%s is an invalid percentage limit" % value)
        return ivalue

    def get_file(value):
        return open(value, 'rb')

    parser = ArgParser(description='A simple plagiarism detection tool for python code')
    parser.add_argument('files', type=get_file, nargs=2,
                        help='the input files')
    parser.add_argument('-l', type=check_line_limit, default=4,
                        help='if AST line of the function >= value then output detail (default: 4)')
    parser.add_argument('-p', type=check_percentage_limit, default=0.5,
                        help='if plagiarism percentage of the function >= value then output detail (default: 0.5)')
    parser.add_argument('-k', '--keep-prints', action='store_true', default=False,
                        help='keep print nodes')
    parser.add_argument('-m', '--module-level', action='store_true', default=False,
                        help='process module level nodes')
    args = parser.parse_args()
    pycode_list = [(f.name, f.read()) for f in args.files]
    try:
        results = detect(
            [c[1] for c in pycode_list],
            keep_prints=args.keep_prints,
            module_level=args.module_level,
        )
    except NoFuncException as ex:
        print('error: can not find functions from {}.'.format(pycode_list[ex.source][0]))
        return

    for index, func_ast_diff_list in results:
        print('ref: {}'.format(pycode_list[0][0]))
        print('candidate: {}'.format(pycode_list[index][0]))
        sum_plagiarism_percent, sum_plagiarism_count, sum_total_count = summarize(func_ast_diff_list)
        print('{:.2f} % ({}/{}) of ref code structure is plagiarized by candidate.'.format(
            sum_plagiarism_percent * 100,
            sum_plagiarism_count,
            sum_total_count,
        ))
        print('candidate function plagiarism details (AST lines >= {} and plagiarism percentage >= {}):'.format(
            args.l,
            args.p,
        ))
        output_count = 0
        for func_diff_info in func_ast_diff_list:
            if len(func_diff_info.info_ref.func_ast_lines) >= args.l and func_diff_info.plagiarism_percent >= args.p:
                output_count = output_count + 1
                print(func_diff_info)

        if output_count == 0:
            print('<empty results>')


if __name__ == '__main__':
    main()
